### PR TITLE
Listen on netlink uevent socket for edev events and send to a socket client

### PR DESCRIPTION
**Pull request by gilgoolon**

### CODE REVIEW

File: client/event_pb2.py
Comment (HorizonRL) [inline client/event_pb2.py:1]: do not upload the auto-generated protobuf file

File: client/main.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client:
+    client.connect(("localhost", 8080))
+
+    while True:
+        data = client.recv(1024)
+
+        event = event_pb2.Event()
+        event.ParseFromString(data)
+
+        print(f"Received Protobuf Event: Action={event.action}, Devname={event.devname}, Vendor={event.vendor}, Model={event.model}, Size={event.size}, Partitions={event.partitions}, Type={event.type}")
```
Comment (HorizonRL) [inline client/main.py:14]: should be wrapped in side a def main function and the main function should be called by if name == main

File: protos/event.proto [hunk 2]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
+
+message Event {
+    string action = 1;
```
Comment (HorizonRL) [inline protos/event.proto:4]: this could be an enum :-)

File: src/block_device.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+constexpr std::string_view BLOCK_DEVICE_INFO_PATH = "/sys/block";
+constexpr std::string_view BLOCK_DEVICE_REMOVABLE_REL_PATH = "removable";
+constexpr std::string_view BLOCK_DEVICE_SIZE_REL_PATH = "size";
+constexpr std::string_view DEV_SYSTEM_PATH = "/dev";
+static constexpr std::string_view DEFAULT_ATTRIBUTE_VALUE = "unknown";
+
+class BlockDevice final
+{
+public:
+    explicit BlockDevice(std::string device_name);
```
Comment (HorizonRL) [inline src/block_device.hpp:15]: const ref

File: src/block_device.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    std::string get_model() const;
+
+    std::string get_vendor() const;
+
+    size_t get_size() const;
+
+    size_t get_partitions_count() const;
+
+    bool is_external() const;
```
Comment (HorizonRL) [inline src/block_device.hpp:25]: in logical oop u have 2 types of getters that often confused by, these are get and retrieve,
the get should only return the value without any added operation while the retrieve can and probably will do something to acquire the needed value, 
the point here is not the naming conventions, but rather then the structure of the block device, it could be like this, the self knows how to retrieve the info or make it a dummy in the pre process (aka a builder will create and inject the found values)

File: src/block_device.hpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#include <string>
```
Comment (HorizonRL) [inline src/block_device.hpp:1]: pragma once

File: src/block_device.hpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+constexpr std::string_view CLASS_BLOCK_DEVICE_INFO_PATH = "/sys/class/block";
+constexpr std::string_view BLOCK_DEVICE_MODEL_REL_PATH = "device/model";
+constexpr std::string_view BLOCK_DEVICE_VENDOR_REL_PATH = "device/vendor";
+constexpr std::string_view BLOCK_DEVICE_INFO_PATH = "/sys/block";
+constexpr std::string_view BLOCK_DEVICE_REMOVABLE_REL_PATH = "removable";
+constexpr std::string_view BLOCK_DEVICE_SIZE_REL_PATH = "size";
+constexpr std::string_view DEV_SYSTEM_PATH = "/dev";
```
Comment (HorizonRL) [inline src/block_device.hpp:9]: Perhaps this should be in a namespace or something responsible for paths

Comment (HorizonRL) [inline src/block_device.hpp:9]: could be static as well

File: src/block_device.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+BlockDevice::BlockDevice(std::string device_name) : _device_name(std::move(device_name)) {}
+
+std::string BlockDevice::get_model() const
+{
+    try
+    {
+        const auto model_path = std::filesystem::path(CLASS_BLOCK_DEVICE_INFO_PATH) / _device_name / BLOCK_DEVICE_MODEL_REL_PATH;
+        return strings::strip(os::read_text_file(model_path), '\n');
+    }
+    catch (...)
```
Comment (HorizonRL) [inline src/block_device.cpp:22]: too broad exception catching

File: src/block_device.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+std::string BlockDevice::get_vendor() const
+{
+    try
+    {
+        const auto vendor_path = std::filesystem::path(CLASS_BLOCK_DEVICE_INFO_PATH) / _device_name / BLOCK_DEVICE_VENDOR_REL_PATH;
+        return strings::strip(os::read_text_file(vendor_path), '\n');
+    }
+    catch (...)
```
Comment (HorizonRL) [inline src/block_device.cpp:35]: too broad exception catching

File: src/block_device.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+size_t BlockDevice::get_size() const
+{
+    try
+    {
+        const auto vendor_path = std::filesystem::path(BLOCK_DEVICE_INFO_PATH) / _device_name / BLOCK_DEVICE_SIZE_REL_PATH;
+        return std::stol(os::read_text_file(vendor_path));
+    }
+    catch (...)
```
Comment (HorizonRL) [inline src/block_device.cpp:48]: too broad exception catching

File: src/block_device.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+bool BlockDevice::is_external() const
+{
+    try
+    {
+        const auto removable_path = std::filesystem::path(BLOCK_DEVICE_INFO_PATH) / _device_name / BLOCK_DEVICE_REMOVABLE_REL_PATH;
+        return std::stoi(os::read_text_file(removable_path));
+    }
+    catch (...)
```
Comment (HorizonRL) [inline src/block_device.cpp:75]: too broad exception catching

File: src/block_device.hpp (hunk 4 above)
Comment (HorizonRL) [inline src/block_device.hpp:25]: u planned this object thinking only about the scenario of printing the result,
meaning when failing to get a value u return a place-catcher (aka "unknown") while this is safe for the user, it's not safe for the logical flow. 
u create a know-all class that is designed to fetch the info and format it (if good or bad),
what if I change the exercise and ask for the program to exit cleanly when finding a device that can not be resolved,
u would have to go into every single logical base of a value fetching and change the flow...
i would throw when an "unknown" is reached and let another being handle the string-ui-readble formating

File: src/common/auto_fd.hpp [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+struct FdCloser final
+{
+    void operator()(const int *fd) const
+    {
+        if (UNIX_INT_ERROR_VALUE != *fd)
+        {
+            ::close(*fd);
+        }
+    }
```
Comment (HorizonRL) [inline src/common/auto_fd.hpp:15]: implement in the cpp

File: src/common/auto_fd.hpp [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+};
+
+using AutoFd = std::unique_ptr<int, FdCloser>;
+
+AutoFd make_auto_fd(int fd);
```
Comment (HorizonRL) [inline src/common/auto_fd.hpp:20]: why is this needed?

File: src/common/auto_fd.cpp [hunk 13]
```diff
@@ -0,0 +1,6 @@
+#include "auto_fd.hpp"
+
+AutoFd make_auto_fd(const int fd)
+{
+    return AutoFd(fd != UNIX_INT_ERROR_VALUE ? new int(fd) : nullptr);
```
Comment (HorizonRL) [inline src/common/auto_fd.cpp:5]: dont use new ;-)

File: src/common/auto_fd.hpp [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+#include "unix.hpp"
+
+struct FdCloser final
+{
+    void operator()(const int *fd) const
+    {
+        if (UNIX_INT_ERROR_VALUE != *fd)
+        {
+            ::close(*fd);
+        }
```
Comment (HorizonRL) [inline src/common/auto_fd.hpp:14]: treat as a dtor, should be wrapped in try and catch

File: src/common/server_socket.cpp [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    server_address.sin_port = ::htons(port);
+    server_address.sin_addr.s_addr = INADDR_ANY;
+    os::covered_call(UNIX_INT_ERROR_VALUE, ::bind, *_socket_fd.get(),
+                     reinterpret_cast<const sockaddr *>(&server_address),
+                     sizeof(server_address));
+}
+
+void ServerSocket::listen(const size_t max_connections) const
+{
+    os::covered_call(UNIX_INT_ERROR_VALUE, ::listen, *_socket_fd.get(),
```
Comment (HorizonRL) [inline src/common/server_socket.cpp:24]: its starting to be a big code base, idk if this approach still fits here

File: src/common/server_socket.cpp [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+                                                 make_auto_fd(os::covered_call(UNIX_INT_ERROR_VALUE, &::socket, AF_INET,
+                                                                               SOCK_STREAM, DEFAULT_NO_FLAGS)))
+{
+    sockaddr_in server_address;
+    server_address.sin_family = AF_INET;
+    server_address.sin_port = ::htons(port);
+    server_address.sin_addr.s_addr = INADDR_ANY;
+    os::covered_call(UNIX_INT_ERROR_VALUE, ::bind, *_socket_fd.get(),
+                     reinterpret_cast<const sockaddr *>(&server_address),
+                     sizeof(server_address));
```
Comment (HorizonRL) [inline src/common/server_socket.cpp:19]: should be a private function, do not directly impl in the ctor

File: src/common/server_socket.hpp [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+#include <netinet/in.h>
+#include <memory>
+
+#include "socket.hpp"
+#include "auto_fd.hpp"
+
+class ServerSocket final
+{
+public:
+    explicit ServerSocket(int port);
```
Comment (HorizonRL) [inline src/common/server_socket.hpp:11]: const and the int size is known, so don't use implicit int decleration

File: src/common/unix.hpp [hunk 18]
```diff
@@ -0,0 +1,6 @@
+#pragma once
+#include <cstddef>
+
+constexpr int UNIX_INT_ERROR_VALUE = -1;
+constexpr int DEFAULT_NO_FLAGS = 0;
+constexpr nullptr_t OPTIONAL_NO_OUTPUT = nullptr;
```
Comment (HorizonRL) [inline src/common/unix.hpp:6]: i prefer to rewrite the flags rather than declaring them here in a global runtime scope, its a big code base, u should prioritize momentary ez to understand rather then preventing ctrl c ctrl v around the code

File: src/event.p.cpp
Comment (HorizonRL) [inline src/event.p.cpp:1]: do not upload the auto-generated protobuf file

File: src/event.p.hpp
Comment (HorizonRL) [inline src/event.p.hpp:1]: do not upload the auto-generated protobuf file

File: src/event.pb.cpp
Comment (HorizonRL) [inline src/event.pb.cpp:1]: do not upload the auto-generated protobuf file

File: src/event.pb.hpp
Comment (HorizonRL) [inline src/event.pb.hpp:1]: do not upload the auto-generated protobuf file

File: src/exceptions.hpp [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+    const char *what() const noexcept override;
+
+private:
+    const int _errnum;
+    const std::string _message;
+};
+
+class DisconnectedException final : public std::exception
+{
+};
```
Comment (HorizonRL) [inline src/exceptions.hpp:22]: rather than creating ur custom class for each exception type (e.g DisconnectedException ), create a base exception that receives an enum for error code (the errno should inherit from it)

Comment (HorizonRL) [inline src/exceptions.hpp:22]: as said before, no need to inherit from std::exception

File: src/main.cpp
Comment (HorizonRL) [inline src/main.cpp:1]: too long main file

File: src/main.cpp [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+std::unique_ptr<Socket> make_netlink_uevent_socket()
+{
+    auto auto_sock = std::make_unique<Socket>(os::covered_call(UNIX_INT_ERROR_VALUE, ::socket, AF_NETLINK, SOCK_RAW, NETLINK_KOBJECT_UEVENT));
+
+    struct sockaddr_nl sa;
+    memset(&sa, 0, sizeof(sa));
+    sa.nl_family = AF_NETLINK;
+    sa.nl_pid = getpid();
+    sa.nl_groups = NETLINK_BROADCAST;
+    os::covered_call(UNIX_INT_ERROR_VALUE, ::bind, auto_sock->get_socket_fd(), reinterpret_cast<sockaddr *>(&sa), sizeof(sa));
```
Comment (HorizonRL) [inline src/main.cpp:31]: why is this code outside of ur API? why does the user need to write this? maybe create a wrapper for this "class"

File: src/main.cpp [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+    return std::move(auto_sock);
+}
+
+std::unique_ptr<Socket> accept_client()
+{
+    ServerSocket server_socket(PORT);
+    server_socket.listen(1);
+    return std::move(server_socket.accept());
```
Comment (HorizonRL) [inline src/main.cpp:40]: why is this code outside of ur API? why does the user need to write this?

File: src/main.cpp [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+        event_to_send.set_action(event.get_action());
+        event_to_send.set_devname(event.get_devname());
+        event_to_send.set_vendor(block_device.get_vendor());
+        event_to_send.set_model(block_device.get_model());
+        event_to_send.set_size(block_device.get_size());
+        event_to_send.set_partitions(block_device.get_partitions_count());
+        event_to_send.set_type(block_device.is_external() ? "external" : "internal");
+        std::string serialized;
+        event_to_send.SerializeToString(&serialized);
+        client_socket->send(serialized);
```
Comment (HorizonRL) [inline src/main.cpp:73]: split this into separate beings and functions

File: src/udevevent.hpp [hunk 23]
```diff
@@ -0,0 +1,22 @@
+#include <string>
+#include <map>
+
+class UDevEvent final
+{
+public:
+    UDevEvent(const std::string_view &event);
```
Comment (HorizonRL) [inline src/udevevent.hpp:7]: explicit

File: src/udevevent.hpp [hunk 24]
```diff
@@ -0,0 +1,22 @@
+#include <string>
```
Comment (HorizonRL) [inline src/udevevent.hpp:1]: pragma once

File: src/udevevent.cpp [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        const auto tokens = strings::split(current, ATTRIBUTE_NAME_VALUE_SEPARATOR, 1); // only split by a single separator
+        if (tokens.size() == 1)                                                         // no separators found
+        {
+            throw std::invalid_argument("Expected " + strings::to_string(ATTRIBUTE_NAME_VALUE_SEPARATOR) + " tokens when splitting event line by '=' token. Line: '" + current + "'.");
+        }
+        const auto attribute_name = tokens.front();
+        const auto attribute_value = tokens.back();
+        _attributes.insert_or_assign(attribute_name, attribute_value);
+    }
```
Comment (HorizonRL) [inline src/udevevent.cpp:36]: should be a named function, do not directly impl inside the ctor

**Review by HorizonRL — COMMENTED** [review]
nice work!

File: src/block_device.hpp (hunk 4 above)
Comment (gilgoolon) [reply src/block_device.hpp:25]: I know. I had it throw an exception since the start but I got lazy in some areas here so skipped writing something prettier…

File: src/block_device.hpp (hunk 3 above)
Comment (gilgoolon) [reply src/block_device.hpp:15]: I did the modern thingy of making a copy and doing std::move(...) in the body...

File: src/main.cpp (hunk 21 above)
Comment (gilgoolon) [reply src/main.cpp:40]: What are you suggesting?

File: src/block_device.hpp (hunk 4 above)
Comment (gilgoolon) [reply src/block_device.hpp:25]: I 100% agree, good catch.

File: src/common/server_socket.cpp (hunk 15 above)
Comment (gilgoolon) [reply src/common/server_socket.cpp:24]: Why though? It's readable enough in my opinion.
What are you suggesting?

File: src/block_device.hpp (hunk 4 above)
Comment (HorizonRL) [reply src/block_device.hpp]: Nah I know u know how to do this but since this is a Rico style assignment I want u to feel the whole experience like it's a for real code

File: src/common/server_socket.cpp (hunk 15 above)
Comment (HorizonRL) [reply src/common/server_socket.cpp]: I do agree that it's kind of a boilerplate, though it's readable for you (the writer) and for me, the reviewer, now let's take some of the future scenarios:

Someone who is not familiar with your code wants to read it, since this is repeatable u could argue that once he figures it out it'll be ez to read ur code, though u r kinda right u r missing the point of the "readable code" principle. 

I'll explain, on a large scale, or even on an ongoing project or most of all collab writing, this approach is kinda problematic, if u were to read the code in it's flow as an English script u are getting the boilerplate onto ur eyes rather then letting ur brain skip and fast-read the wrap around boilerplate (e.g the repeated return codes)

Think of it this way ocp will break soon enough when I'll give u a command that won't follow ur calling convention, then u will break lsp to change it

I would argue that if u still want to take that approach u would need to wrap each of the "raw" calling to this wrapper function with a specific named function for clear reading and maintenance

File: src/main.cpp (hunk 21 above)
Comment (HorizonRL) [reply src/main.cpp]: Make the ctor call listen by itself 
A server sock is not defined (logically) when it's not bound nor listening

**Comment by gilgoolon** [comment]
I've left the pythonic client as-is since I'm going to spend more time on it on the second part of the task - so please don't comment on it xD

Besides the comments you left, I refactored the main file (and some additional functions) to a more SOLID approach, since I noticed this is starting to be a bigger codebase and I was lazy at the beginning (and didnt anticipate it)
